Optimization coding pass over AudioMixerSlavePool and AvatarMixerSlavePool (again!)

odysseus654 opened the pull request:
This is an optimization pass over the mutex logic in AudioMixerSlavePool and AvatarMixerSlavePool (which is effectively the same code).

This is a continuation of the changes in #1358 , with more of a focus on contracts and invariants and effectively no use of mutexes.  This should (almost by definition) have reduced lock contention, but "thread-safety through the use of code comments" may have negative side effects.  I don't know which of these PRs will end up being used (perhaps a mixture of both?)

Additional changes made to this PR:
- Removed all use of "friend" by factoring out shared variables into a private class that the slaves can see but pool users cannot
- A number of counters have been changed from "int" to "std::atomic<int>"
- Much documentation of invariants and code contracts

Testing is almost identical to #1358 :
- Smoketesting: make sure that audio and avatars still work on a server built with this (requires a server with multiple users)
- Race conditions: exercise a server long enough to know whether there are any lockups or crashes that occur in these two areas
- Profile: see if this improves the profiling numbers that have been described in issue High CPU usage of assignment client when idle on Linux #1355

odysseus654 commented:
Hey, if anyone wants to use std::shared_mutex, there's a polyfill in the commits (that was later yanked out) ;-)

HifiExperiments commented on assignment-client/src/avatars/AvatarMixerSlavePool.cpp:
```diff
@@ -75,10 +77,10 @@ void AvatarMixerSlavePool::broadcastAvatarData(ConstIter begin, ConstIter end,
         slave.configureBroadcast(begin, end, lastFrameTimestamp, maxKbpsPerNode, throttlingRatio,
-            _priorityReservedFraction);
-   };
+                                    _priorityReservedFraction);
```
indentation?

odysseus654 replied:
I don't know whether I should copy/paste it again to have it aligned to the start of "begin" or force it to go back to where it was really.

HifiExperiments replied:
oh wait sorry this an argument not the body of a lambda, nvm me this looks correct (sorry reviewing on my phone)

HifiExperiments commented on assignment-client/src/avatars/AvatarMixerSlavePool.h:
```diff
@@ -34,30 +73,26 @@ class AvatarMixerSlaveThread : public QThread, public AvatarMixerSlave {
-class AvatarMixerSlavePool {
-    using Queue = tbb::concurrent_queue<SharedNodePointer>;
-    using Mutex = std::mutex;
+class AvatarMixerSlavePool : private AvatarMixerSlavePoolData {
```
I like this approach, I think it’s a lot cleaner and easier to follow.  from a naming perspective, it seems weird for a Pool to be a PoolData; I’d expect a Pool to have a PoolData maybe?  or a different name for PoolData?

odysseus654 replied:
Heh, suggestions welcome?  I could stick this in as a private field too which could work, would just require variables to be qualified when accessed from the pool

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-ba09d2a-_536b1ed1.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ba09d2a-_536b1ed1.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ba09d2a-_536b1ed1.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ba09d2a-.tar_536b1ed1.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ba09d2a-.tar_536b1ed1.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ba09d2a-_536b1ed1.exe

odysseus654 commented:
Made suggested changes to this PR:
- changed "slave" to "worker" when dealing with classes in this file
- moved the "data" struct to be a private member instead of a base class

Attempted to factor out common code to a template base class but... can't do that and still use Qt

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2dff102-.tar_aa2c3074.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2dff102-_aa2c3074.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2dff102-.tar_aa2c3074.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2dff102-_aa2c3074.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2dff102-_aa2c3074.exe

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-2dff102-_aa2c3074.dmg

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5c42466-.tar_569afb18.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5c42466-.tar_569afb18.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5c42466-_569afb18.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5c42466-_569afb18.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5c42466-_569afb18.exe

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-5c42466-_569afb18.dmg

HifiExperiments commented on assignment-client/src/audio/AudioMixerSlavePool.h:
```diff
@@ -22,39 +22,75 @@
     bool try_pop(SharedNodePointer& node);
 
-    AudioMixerSlavePool& _pool;
+    AudioMixerWorkerPoolData& _pool;
```
maybe _poolData?

odysseus654 replied:
Whatever it is should be short, it's _data in the pool class and _pool here, and is a qualifier in front of all field accesses.

HifiExperiments replied:
I’m fine with it being _data in both places

HifiExperiments commented on assignment-client/src/audio/AudioMixerSlavePool.h:
```diff
@@ -63,49 +99,34 @@ class AudioMixerSlavePool {
+    AudioMixerWorkerPoolData _data;
 };
 
 #endif // hifi_AudioMixerSlavePool_h
```
worker

odysseus654 replied:
We may want to name this the same thing that we name the value above so both classes use the same value in accessing the data.

HifiExperiments replied:
sorry I meant the comment here, looks like there’s a few comments that still use “slave” instead of “worker”.  but I understand wanting to keep the changes local to this file so not changing everywhere right now, we can do it in pieces later

odysseus654 replied:
It involves a file renaming if we wanna go all the way on that, plus renaming AudioMixerSlave

HifiExperiments replied:
let’s save it for a larger pass, I’ve been meaning to do this for a while, I appreciate your making the changes here

HifiExperiments commented on assignment-client/src/avatars/AvatarMixerSlavePool.cpp:
```diff
@@ -151,53 +158,61 @@ void AvatarMixerSlavePool::setNumThreads(int numThreads) {
+            {
+                Lock poolLock(_data.poolMutex);
+                if (_data.numFinished < _data.numThreads) {
+                    _data.poolCondition.wait(poolLock, [&] {
```
hm…is it possible that all the threads will finish and notify between line 188 above and here?  so this wait would never finish?

odysseus654 replied:
Heh, I thought about that this morning.  The most recent commit now has all "condition.notify" calls holding the corresponding lock when calling it (thus forcing the notify to be before line 192 or after line 194) and the "if" statement wrapping all "condition.wait" calls bypasses it if the conditions for it to be notified already exist.

odysseus654 commented:
Added commit: renamed _pool to _data as suggested.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-3dd0fb4-_bc5d9de4.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-3dd0fb4-_bc5d9de4.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-3dd0fb4-.tar_bc5d9de4.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-3dd0fb4-_bc5d9de4.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-3dd0fb4-.tar_bc5d9de4.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-3dd0fb4-_bc5d9de4.exe

HifiExperiments reviewed (approved):
looks good to me, definitely want to stress test this with a bunch of people on a server and do another perf analysis to see if this made a difference

ksuprynowicz commented:
Linux build of the current version fails for me with following error:
```
/home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:216:37: error: ‘_slaves’ was not declared in this scope; did you mean ‘_xsaves’?
  216 |     assert(_data.numThreads == (int)_slaves.size());
```

odysseus654 commented:
Gack, was worried something like that might happen.  I haven't been using debugging builds so assert() is meaningless to me.  I'll fix that.

daleglass commented:
Might be worth trying to test this with #1149

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-12dc362-_45d33127.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-12dc362-_45d33127.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-12dc362-.tar_45d33127.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-12dc362-_45d33127.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-12dc362-.tar_45d33127.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-12dc362-_45d33127.exe

ksuprynowicz commented:
Current version builds well but assignment client child threads continuously crash with message:
```
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:44: AudioMixerWorkerThread::wait()::<lambda()>: Assertion `_data.numStarted <= _data.numThreads' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:58: void AudioMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads && _data.numFinished <= _data.numStarted' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:58: void AudioMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads && _data.numFinished <= _data.numStarted' failed.
[09/25 23:33:28] [CRITICAL] [default] [26583] [assignment-client-monitor] Child process 27216 on port 0has crashed with exit code 6. Removed from internal map.
```

odysseus654 commented:
I read this report as "asserts are good" and "next time smoketest the darned thing".  Will look, hopefully a quick fix.

odysseus654 commented:
I'm unable to smoketest this on my windows machine, so made changes in the code based directly on your printed error messages.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5686023-_3a4e8a7f.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-5686023-_3a4e8a7f.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5686023-_3a4e8a7f.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5686023-.tar_3a4e8a7f.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5686023-.tar_3a4e8a7f.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-5686023-_3a4e8a7f.exe

ksuprynowicz commented:
The current version crashes with a different message. From domain server side:
```
[09/26 09:46:15] [DEBUG] [hifi.domain_server] Haven't heard from  213.134.187.199:24938 ""  in  11895  msec
[09/26 09:46:16] [DEBUG] [hifi.networking.ice] Removing silent node "Avatar Mixer" (W) {0c89c76d-aa3c-4a02-bc92-543f4e5bbb91}(53345) 213.134.187.199:24938 / 192.168.0.15:56776
```
and from assignment client side:
```
[09/26 09:46:03] [WARNING] [default] [54542] [avatar-mixer] Setting::Interface::init() for key "private/selectedMetaverseURL" - Manager not yet created. Settings persistence disabled.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:54: void AvatarMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:54: void AvatarMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads' failed.
[09/26 09:46:03] [CRITICAL] [default] [54378] [assignment-client-monitor] Child process 54542 on port 0has crashed with exit code 6. Removed from internal map.
[09/26 09:46:04] [DEBUG] [hifi.networking] [54383] [entity-server] LimitedNodeList::sendPacket called without active socket for node "Avatar Mixer" (W) {0c89c76d-aa3c-4a02-bc92-543f4e5bbb91}(53345) 213.134.187.199:24938 / 192.168.0.15:56776 - not sending
[09/26 09:46:04] [DEBUG] [hifi.networking] [54385] [entity-script-server] LimitedNodeList::sendPacket called without active socket for node "Avatar Mixer" (W) {0c89c76d-aa3c-4a02-bc92-543f4e5bbb91}(53345) 213.134.187.199:24938 / 192.168.0.15:56776 - not sending
```

ksuprynowicz commented:
It seems that audio thread crashes too, with same message as last time:
```
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:58: void AudioMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:58: void AudioMixerWorkerThread::notify(bool): Assertion `_data.numFinished < _data.numThreads' failed.
[09/26 09:52:02] [CRITICAL] [default] [54699] [assignment-client-monitor] Child process 54867 on port 0has crashed with exit code 6. Removed from internal map.
```

odysseus654 commented:
Thank you for your efforts on this.  I can't guarantee that this won't be the last error (likely will be at least a few more, considering the lack of smoketesting)

I've made another commit to ensure that newly created workers are properly waiting (and not active) before leaving "resize"

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2611422-_83f2b526.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2611422-_83f2b526.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2611422-.tar_83f2b526.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2611422-.tar_83f2b526.gz

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-2611422-_83f2b526.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-2611422-_83f2b526.exe

ksuprynowicz commented:
I just gave a try for the most recent build, and it crashes in a different way now. Audio mixer crashes this way:
```
[09/27 20:55:39] [DEBUG] [hifi.assignment-client] [111780] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/27 20:55:39] [DEBUG] [default] [111780] [<audio-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:56: void AudioMixerWorkerThread::wait(): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:56: void AudioMixerWorkerThread::wait(): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:56: void AudioMixerWorkerThread::wait(): Assertion `_function' failed.
[09/27 20:55:39] [CRITICAL] [default] [111590] [assignment-client-monitor] Child process 111780 on port 0has crashed with exit code 6. Removed from internal map.

```
And avatar mixer crashes this way:

```
[09/27 20:57:09] [DEBUG] [hifi.assignment-client] [111884] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/27 20:57:09] [DEBUG] [default] [111884] [<avatar-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:52: void AvatarMixerWorkerThread::wait(): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:52: void AvatarMixerWorkerThread::wait(): Assertion `_function' failed.
[09/27 20:57:09] [CRITICAL] [default] [111590] [assignment-client-monitor] Child process 111884 on port 0has crashed with exit code 6. Removed from internal map.

```

ksuprynowicz commented:
> Thank you for your efforts on this. I can't guarantee that this won't be the last error (likely will be at least a few more, considering the lack of smoketesting)

I'll be happy to test it as many times as needed :)
Thank you for the work you are putting into it.

odysseus654 commented:
Heh, plus those asserts you got were just added yesterday; was suspecting this might be the underlying causes of the crashes (rather than just "misbehaviors").

odysseus654 commented:
Okay did some code re-ordering w.r.t. worker thread startup, assuming that these crashes are the first time through the loop.

Less likely chance of this crash but no-crash doesn't mean it works; I'm really wanting to not see *any* assert()s in this.

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-fd90b96-.tar_df59a42c.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-fd90b96-_df59a42c.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-fd90b96-.tar_df59a42c.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-fd90b96-_df59a42c.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-fd90b96-_df59a42c.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-fd90b96-_df59a42c.exe

ksuprynowicz commented:
It crashes with both assertions now:
```
[09/28 09:11:59] [DEBUG] [default] [36663] [<avatar-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:59: void AvatarMixerWorkerThread::wait(): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:59: void AvatarMixerWorkerThread::wait(): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:18: virtual void AvatarMixerWorkerThread::run(): Assertion `_data.numStarted < _data.numThreads' failed.
[09/28 09:11:59] [CRITICAL] [default] [36202] [assignment-client-monitor] Child process 36663 on port 0has crashed with exit code 6. Removed from internal map.
```
```
[09/28 09:13:29] [DEBUG] [default] [36724] [<audio-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:63: void AudioMixerWorkerThread::wait(): Assertion `_function' failed.
[09/28 09:13:29] [CRITICAL] [default] [36202] [assignment-client-monitor] Child process 36724 on port 0has crashed with exit code 6. Removed from internal map.
```

odysseus654 commented:
okay dug into "exit code 6" and it looks like this crash basically happens the moment an assert() failure occurs.

Still looks like we're having problems with new worker startup (I think), I'll walk through that chain of logic a bit further.

odysseus654 commented:
Okay new changes (as described in the commit):
- New threads now unconditionally enter a wait without checking if numStarted==numThreads (as numThreads is changing out-of-band in this case.  Yes it's violating the contract in the code comments, the worker thread is "active" without "owning" the state.  Hopefully not too many more headaches with this)
- Moved numStarted++ inside the workerLock so as to avoid spurious worker thread wakes

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-56fe8f8-_5389d3fd.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-56fe8f8-_5389d3fd.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-56fe8f8-.tar_5389d3fd.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-56fe8f8-.tar_5389d3fd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-56fe8f8-_5389d3fd.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-56fe8f8-_5389d3fd.exe

ksuprynowicz commented:
Right now it crashes with a message:
```
[09/28 16:29:52] [DEBUG] [hifi.assignment-client] [113656] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/28 16:29:52] [DEBUG] [default] [113656] [<audio-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:65: void AudioMixerWorkerThread::wait(bool): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:65: void AudioMixerWorkerThread::wait(bool): Assertion `_function' failed.
[09/28 16:29:52] [CRITICAL] [default] [113574] [assignment-client-monitor] Child process 113656 on port 0has crashed with exit code 6. Removed from internal map.
```
and
```
[09/28 16:31:22] [DEBUG] [hifi.assignment-client] [113725] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/28 16:31:22] [DEBUG] [default] [113725] [<avatar-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:61: void AvatarMixerWorkerThread::wait(bool): Assertion `_function' failed.
[09/28 16:31:22] [CRITICAL] [default] [113574] [assignment-client-monitor] Child process 113725 on port 0has crashed with exit code 6. Removed from internal map.
```

odysseus654 commented:
Starting to think I may not know how condition_var.wait() works.  Adding a mutex blocker to prevent the wait from occurring before all the threads have started up.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ff873bd-_ecba866a.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-ff873bd-_ecba866a.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ff873bd-_ecba866a.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ff873bd-.tar_ecba866a.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ff873bd-.tar_ecba866a.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-ff873bd-_ecba866a.exe

ksuprynowicz commented:
> Starting to think I may not know how condition_var.wait() works. Adding a mutex blocker to prevent the wait from occurring before all the threads have started up.

It still crashes, the current messages are:

[```
09/30 23:26:38] [DEBUG] [hifi.assignment-client] [90296] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/30 23:26:38] [DEBUG] [default] [90296] [<avatar-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:61: void AvatarMixerWorkerThread::wait(bool): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/avatars/AvatarMixerSlavePool.cpp:18: virtual void AvatarMixerWorkerThread::run(): Assertion `_data.numStarted < _data.numThreads' failed.
[09/30 23:26:38] [CRITICAL] [default] [90045] [assignment-client-monitor] Child process 90296 on port 0has crashed with exit code 6. Removed from internal map.
```

```
[09/30 23:26:08] [DEBUG] [hifi.assignment-client] [90207] [assignment-client] Received a PacketType::CreateAssignment - attempting to unpack.
[09/30 23:26:08] [DEBUG] [default] [90207] [<audio-mixer>] resize: set 40 threads (was 0)
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:65: void AudioMixerWorkerThread::wait(bool): Assertion `_function' failed.
assignment-client: /home/ksuprynowicz/Vircadia/source/assignment-client/src/audio/AudioMixerSlavePool.cpp:22: virtual void AudioMixerWorkerThread::run(): Assertion `_data.numStarted < _data.numThreads' failed.
[09/30 23:26:08] [CRITICAL] [default] [90045] [assignment-client-monitor] Child process 90207 on port 0has crashed with exit code 6. Removed from internal map.
```

odysseus654 commented:
I managed to piece together a linux debugging environment and debugged it enough to no longer assert (note to self: deleting assert commands is a great way to remove those pesky asserts).

Anyhow, really hoping this is the last patch necessary in order to successfully smoketest this.  Also hoping to have a chance to flatten these test commits (there's an awful lot of them) although I understand if that doesn't happen.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-555443a-_f7e61b8a.dmg

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-_f7e61b8a.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-.tar_f7e61b8a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-.tar_f7e61b8a.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-_f7e61b8a.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-_f7e61b8a.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-.tar_f7e61b8a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-.tar_f7e61b8a.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-555443a-_f7e61b8a.sh

ksuprynowicz commented:
I just gave it a try, and nothing is crashing right now, I'm able to connect, but only audio mixer starts. Other types of assignment clients don't start.

ksuprynowicz commented:
I'm sorry. I ran assignment server with wrong command line parameters. Everything seems to work correctly now, but CPU usage is still very high even when no client is connected.

odysseus654 commented:
Heh, finally got it! So I guess the question is what the profiler is saying about this now.

ksuprynowicz commented:
Congratulations!
I will do a profiler run today when I finish work and see what is shows :)

ksuprynowicz commented:
CPU use seems to be significantly lower now compared to master when built with the same compiler flags - it drops almost 2 times when no user is logged in.
Here are the profiling results for audio mixer:

```
-   83.79%     0.55%  AudioMixerWorke  [kernel.kallsyms]            [k] entry_SYSCALL_64_after_hwframe
   - 83.24% entry_SYSCALL_64_after_hwframe
      - do_syscall_64
         - 81.87% __x64_sys_futex
            - do_futex
               - 60.03% futex_wait
                  - 44.78% futex_wait_queue_me
                     - 43.66% schedule
                        - __schedule
                           - 27.91% pick_next_task_fair
                              - 27.66% newidle_balance
                                 - 24.60% load_balance
                                    - 16.37% find_busiest_group
                                       - update_sd_lb_stats.constprop.0
                                          - 2.43% cpumask_next_and
                                               _find_next_bit
                                            1.51% idle_cpu
                                    + 5.40% raw_spin_rq_lock_nested
                                   1.18% update_blocked_averages
                           - 6.39% finish_task_switch.isra.0
                              - 5.11% __perf_event_task_sched_in
                                 + 3.33% ctx_sched_in
                                 + 0.59% perf_event_sched_in
                           + 3.25% dequeue_task_fair
                           + 2.46% psi_task_switch
                           + 1.04% __perf_event_task_sched_out
                       0.74% plist_add
                  + 14.39% futex_wait_setup
               + 20.80% futex_wake
         + 1.04% syscall_exit_to_user_mode

```
And for avatar mixer:

```
-   82.22%     0.52%  AvatarMixerWork  [kernel.kallsyms]        [k] entry_SYSCALL_64_after_hwframe
   - 81.69% entry_SYSCALL_64_after_hwframe
      - do_syscall_64
         - 80.28% __x64_sys_futex
            - do_futex
               - 55.44% futex_wait
                  + 36.85% futex_wait_queue_me
                  + 17.69% futex_wait_setup
               - 23.86% futex_wake
                  + 16.12% _raw_spin_lock
                  + 5.26% wake_up_q
                  + 0.96% mark_wake_futex
         + 1.01% syscall_exit_to_user_mode
```

ksuprynowicz commented:
Audio mixer uses about 32% of a core right now when no one is logged in, and avatar mixer uses about 15% with compiler flags as in PR 1424. I suggest pulling changes from master to make comparisons easier.

ksuprynowicz commented:
The weird thing is that high CPU usage disappears completely on PR 1429. I have no idea why.

ksuprynowicz commented:
We tested it with Kalila and PR 1429 indeed seems to solve high CPU usage completely. It dropped to 3% for audio mixer and 3% for avatar mixer when nobody is logged in.

odysseus654 commented:
Okay kewl! I'm gonna want to flatten all those commits I made while testing this (because ew), but other than that it sounds like we've cleared the "at least as good as master" bar.

odysseus654 commented:
Okay I looked at the history and... it may not be as bad as I thought it was.  I'll squish it if it makes sense to others, otherwise I'll say this is good to merge.

HifiExperiments commented:
sorry if I’m missing it in the above comments: did this PR have a measurable difference then?  if this isn’t an improvement, maybe we don’t need to merge it?

ksuprynowicz commented:
> sorry if I’m missing it in the above comments: did this PR have a measurable difference then? if this isn’t an improvement, maybe we don’t need to merge it?

With current version it does have a measurable improvement, but I suggest to wait with testing it until PR 1429 is merged. I will benchmark it again then and see if it also improves things or not.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-_1f9347d5.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-.tar_1f9347d5.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-_1f9347d5.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-.tar_1f9347d5.gz

**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-4848d86-_1f9347d5.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-_1f9347d5.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-.tar_1f9347d5.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-.tar_1f9347d5.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-4848d86-_1f9347d5.sh

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-Interface-PR1361-8397fd6-_20df505e.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-8397fd6-_20df505e.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-8397fd6-.tar_20df505e.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-8397fd6-.tar_20df505e.Z

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-8397fd6-_20df505e.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1361/Vircadia-PR1361-8397fd6-_20df505e.exe

digisomni commented:
Needs another QA pass to confirm whether or not it performs better than current master.

namark commented:
I tried to test audio mixer locally, just idling as suggested, and couldn't really find any significant performance difference between this and the master branch. Here is what the CPU utilization looks like for me:
![plot1](https://user-images.githubusercontent.com/5012637/152611379-6b5ecdda-8258-43b9-8abf-22e57e82f761.png)
![plot2](https://user-images.githubusercontent.com/5012637/152611383-89006956-1f5c-407d-8bae-1ac83364aea0.png)

I tried a profiler as well, and it does seem that this version spends a bit less time in the run and wait functions of the two classes, but it's just by a few percent and not very consistent between runs, so nothing conclusive there either.

Since the high CPU usage when idle is fixed now on master, I guess it would make more sense to try some sort of a stress test, but I don't fully understand the changes so not sure what would be a good way to do it. I could maybe try using the web sdk to connect a bunch of dummy clients and send some random audio samples, if that's something we expect to be improved here.
